src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java

```diff
@@ -83,15 +88,21 @@
 
         try {
             return RetryUtils.runWithRetry(s3ClientExceptionRetryConfig, () -> {
+                long downloaded = 0;
                 try (InputStream inputStream = regionClient.getObject(getObjectRequest)) {
-                    long downloaded = download(inputStream, messageDigest);
+                    downloaded = download(inputStream, messageDigest);
                     if (downloaded == 0) {
                         // If 0 byte is read, it's fairly certain that the inputStream is closed.
                         // Therefore throw IOException to trigger the retry logic.
                         throw new IOException("Failed to read any byte from the inputStream");
                     } else {
                         return downloaded;
                     }
+                } catch (S3Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)